concepts: basic concept (#450)
exercism/lua: concepts/basics/about.md

# About

[Lua](https://lua.org) is a simple yet powerful, efficient, lightweight, embeddable scripting language. It is designed, implemented, and maintained at [PUC-RIO](https://www.puc-rio.br/) in Brazil since 1993 by the team of Roberto Ierusalimschy, Waldemar Celes and Luiz Henrique de Figueiredo.

## Comments

> Singleline comments in Lua start with `--` and multiline comments start with `--[[` and end with `]]`.

Reviewer: ```suggestion
Single line comments in Lua start with `--` and multiline comments start with `--[[` and end with `]]`.
```

As far as I can tell, it's either "single line" or "single-line" but "multiline" or "multi-line".

```lua
-- This is a single line comment.

--[[This is
a multiline
comment
]]
```

## Variables

There are three types of variables in Lua: global variables, local variables and table fields. To keep things simple only local variables are explained here. By default all variables are global so we have to declare local variables as such with the keyword `local`.
```lua
local x
```

A value can be assigned to a variable by using the `=` operator.
```lua
local x = 1
```

The scope of a `local` variable is limited to the block where it is declared. A block can be the body of a control structure, the body of a function or a file or string where the variable is declared.
```lua
local x = 1

> if x == 1 then
>     local x = 2
>     print(x) -- prints 2
> end

Reviewer: ```suggestion
if x == 1 then
  local x = 2
  print(x) -- prints 2
end
```

We use 2 line indent in the exercises so we should probably use 2 line indents here as well

print(x) -- prints 1
```

## Functions

> A function is declared with the keyword `function` and, like variables, is global by default too. A function definition has a name, a list of parameters and a body. The functions exits when the keyword `return` is reached and returns all values listed after it. Without a specified `return`, the function will return `nil`.

Reviewer: ```suggestion
A function is declared with the keyword `function` and, like variables, is global by default. A function definition has a name, a list of parameters and a body. The functions exits when the keyword `return` is reached and returns all values listed after it. Without a specified `return`, the function will return `nil`.
```

The "too" seems redundant to me because of the "like variables"

```lua
-- Defining a local function which returns "Hello " followed by 
-- the name which was given as parameter
local function say_hello(name)
    return 'Hello ' .. name
end
```

## Tables

Tables are the only data structuring mechanism in Lua and very powerful. They are used to represent arrays, sets, object, modules and many more. Essentially a table in Lua is an associative array.
```lua
-- create a table and assign its reference to tbl
local tbl = {}

-- new entry with key "x" and value 1
tbl['x'] = 1

-- printing the value
print(tbl['x'])  --> 1

> -- not initialized table fields evalute to nil

Reviewer: ```suggestion
-- uninitialized table fields evaluate to nil
```

print(tbl['y'])  --> nil

-- there is also syntactic sugar - the following is the same as above
tbl.y = 1
print(tbl.y)    --> 1
```

This is just a tiny insight which is needed for the next section. We come back to this topic is detail later.

## Modules

A module is a collection of code which is stored in a `.lua` source file. Everything that the modules exports is stored in a table and this table is returned at the end of the file. So the table basically acts as a kind of namespace.
```lua
-- say-hello.lua
local say = {}

function say.hello()
    print('Hello!')
end

return say
```

> That module can then be loaded into another codebase through the function `require`, that returns the table.

Reviewer: ```suggestion
That module can then be loaded into another file through the function `require`, that returns the table.
```

"Codebase" implies to me that it's a separate project so I think "file" fits better.
